Fix bug with static variables in two activity models

hendriko373 opened the pull request:
This PR intents to fix a bug related to cache invalidation in the Duan-Sun and Rumpf activity models. In these models some variables where declared as static, but never updated. As a result, the cached values are not invalidated between sequential computations that make use of the activity models.

In a first attempt the variables were made non-static, resolving the bug. Since this gives up caching altogether, some efficiency might be lost-- for a specific use-case I noted a runtime that took two times longer. Therefore, caching for the variables is reintroduced, with the caching object created and stored next to the `charged` species list that it is supposed to be in sync with.

@allanleal Being a first-time contributor here, I might easily have missed better options. I am all ears for suggestions.

allanleal commented:
Hi @hendriko373 , I see the issue. The assumption was that these models would always be called after another that exported the `AqueousMixture` object. Perhaps a simpler fix would be moving the block of code which determines the species indices to the parent scope, above the lambda function, and then capture them by copy. This would require creating an auxiliary `AqueousMixture` object.

hendriko373 commented:
> Hi @hendriko373 , I see the issue. The assumption was that these models would always be called after another that exported the `AqueousMixture` object. Perhaps a simpler fix would be moving the block of code which determines the species indices to the parent scope, above the lambda function, and then capture them by copy. This would require creating an auxiliary `AqueousMixture` object.

Hi Allan. Thanks for the suggestion. Is it possible that the auxiliary mixture and the one referenced to in the `ActivityPropsRef` object, passed to the `fn` lambda, get out of sync? If not, I am happy to adapt this PR with your suggestion.

hendriko373 committed:
Fix bug with static variables in activity models

Some variables where declared static in the Duan-Sun and Rumpf activity
models, while they where never invalidated. Therefore, running multiple
calculations in sequence for different chemical systems and states, led
to a misuse of cached values.

An auxiliary `AqueousMixture` object is created ouside of the lambda and
the precalculated indices are captured by value to retain the original
caching intent.

hendriko373 commented:
@allanleal I updated the branch, using the auxiliary mixture object, as you suggested.

hendriko373 commented:
@allanleal In case there is anything I can help with to further this PR, please let me know.

allanleal commented:
~Hi Hendrik, thanks for reaching out. I'll try to revise this tomorrow morning, before I'm away until Thursday.~
It looks good. Once the ci checks below pass, I'll merge.

hendriko373 commented:
> ~Hi Hendrik, thanks for reaching out. I'll try to revise this tomorrow morning, before I'm away until Thursday.~ It looks good. Once the ci checks below pass, I'll merge.

Great, thanks!
